### PR TITLE
Added `dvclive.set_step`

### LINKED ISSUE #144: Custom step
Extracted from https://github.com/iterative/dvclive/pull/142#pullrequestreview-738942719.

- What is the intended use case for a custom step?
- Should it overwrite existing values for the same step (it doesn't)?
- Should results be ordered by write time or step (they are ordered by write time)?
- Why set custom steps at the metric level with dvclive.log(step=n) since the step value should probably apply to all metrics?
- If I log one metric and then set a different step for a second metric, which step number should be used for the first metric (it will have a different step in the tsv and the summary json)?

### LINKED ISSUE #154: Add `dvclive.set_step`
From #144. Remove `step` from `dvclive.log` level and add a new method `dvclive.set_step` for enabling "custom step" behavior.

### PR DESCRIPTION
**Pull request by daavoo**
* [X] ❗ I have followed the [Contributing to DVCLive](https://github.com/iterative/dvclive/blob/master/CONTRIBUTING.md) guide.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Closes #154 
Closes #144 (Not sure if actually closes @dberenbaum ?)

`dvc.org` P.R: https://github.com/iterative/dvc.org/pull/2823

---

- Removed `step` arg from `dvclive.log`
- Added new method `dvclive.set_step` 
- Added `MetricLogger.get_step` and `MetricLogger.set_step` per https://github.com/iterative/dvclive/pull/142#discussion_r703260055

`dvclive.set_step` implicitly calls `next_step`.  Jump to https://github.com/iterative/dvclive/pull/157#discussion_r706425968 for explanation

### CODE REVIEW

File: dvclive/metrics.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ def __init__(
         self._step: int = 0
         self._html: bool = html
         self._summary = summary
-        self._metrics: Dict[str, float] = OrderedDict()
+        self._metrics: Dict[str, Tuple[float, float]] = OrderedDict()
```
Comment (dberenbaum) [inline dvclive/metrics.py:32]: What's the reason for this change?

**Review by dberenbaum — APPROVED** [review]

File: dvclive/metrics.py [hunk 2]
```diff
@@ -100,10 +100,14 @@ def summary_path(self):
...
         return self.dir + "_dvc_plots/index.html"
 
-    @property
-    def step(self):
+    def get_step(self) -> int:
         return self._step
 
+    def set_step(self, step: int):
+        if self._metrics:
+            self.next_step()
```
[the remark this answers is not in the document]
Comment (daavoo) [reply dvclive/metrics.py:108]: I added this call mainly to preserve the behavior of a (probably corner case) scenario regarding the implicit `next_step` (discussed in #156 ). TBH the implicit `next_step `is not really documented and so this scenario would be difficult to reach.

Given the following code (originally included in tests):

```python
    steps = [0, 62, 1000]
    metrics = [0.9, 0.8, 0.7]

    for step, metric in zip(steps, metrics):
        dvclive.log("m", metric, step=step)
```

The latest summary (`latest.json`) generated would contain:

```json
{
    "step": 62,
    "m": 0.8
}
```

However, when removing `step` from `dvclive.log` and adding `dvclive.set_step`:

```python
    steps = [0, 62, 1000]
    metrics = [0.9, 0.8, 0.7]

    for step, metric in zip(steps, metrics):
        dvclive.set_step(step)
        dvclive.log("m", metric)
```


The latest summary (`latest.json`) generated would contain:

```json
{
    "step": 1000,
    "m": 0.8
}
```

Which is incorrect.

As I said, this is basically due to the `next_step` side effect in `dvclive.log`, which we could remove if we come to an angreement in #156 . 

The following code would work as intended and the `next_step` inside `set_step` will be just skipped (due to `if self._metrics: `) so the scenario you propose would behave as intended if the user calls `dvclive.next_step` manually:

```python
    steps = [0, 62, 1000]
    metrics = [0.9, 0.8, 0.7]

    for step, metric in zip(steps, metrics):
        dvclive.set_step(step)
        dvclive.log("m", metric)
        dvclive.next_step()
```

File: dvclive/metrics.py (hunk 1 above)
Comment (daavoo) [reply dvclive/metrics.py:32]: It shouldn't had been pushed. I was considering altering `self._metrics` structure to handle the implicit `next_step` call in `dvclive.log`.

Comment (dberenbaum) [reply dvclive/metrics.py]: Is `Dict[str, float]` always correct? I thought that `self._metrics` could be a nested dict?

File: dvclive/metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dberenbaum) [reply dvclive/metrics.py:108]: Interesting. I think it makes sense to avoid asking user to call both `next_step` and `set_step` together. I also think the `{"step": 62, "m": 0.8}` summary would be more expected for users in case they forget to end their code with a call to `next_step`. Maybe we should be writing out the summary at each `log` call?

@pared Was there a reason for this original test to skip the summary for the last step?

File: dvclive/metrics.py (hunk 1 above)
Comment (daavoo) [reply dvclive/metrics.py]: Yep, the typing is not correct. Recursive type hint is kind of a debatable topic. In addition, we should be probably using `typing.OrderedDict` . I guess `OrderedDict[str, Any]` would be the "proper" way?

File: dvclive/metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dberenbaum) [reply dvclive/metrics.py:108]: I discussed with @pared, and it seems like we may need to write out tsv and summary results at each `dvclive.log()` call so that:
1. `next_step` and `set_step` are consistent.
2. Only one of them is needed.
3. Behavior is consistent regardless of whether they are called at the beginning or end.

[the remark this answers is not in the document]
Comment (daavoo) [reply dvclive/metrics.py:108]: Should that change be implemented before the one added in this P.R.?

[the remark this answers is not in the document]
Comment (pared) [reply dvclive/metrics.py:108]: @dberenbaum main reason was to avoid too many I/O operations, but now it seems that current behaviour is standing in a way of correct behaviour of whole library, so I would not worry about it for now.

> Should that change be implemented before the one added in this P.R.?

I don't think it matters, as long as we make a release after both are incorporated to master.

[the remark this answers is not in the document]
Comment (daavoo) [reply dvclive/metrics.py:108]: @dberenbaum @pared even if we consider the`summary` behavior fixed (by moving it to `log`), it doesn't feel like there is a way to fulfill the 3 points listed above when considering `html` and `checkpoints`.

[the remark this answers is not in the document]
Comment (daavoo) [reply dvclive/metrics.py:108]: I think we should consider merging this P.R., if there are not any changes regarding the initial scope (remove `step` from `log` level,  and address the other changes raised during comments in separated issues/pr (i.e. #162 , #145 ). Does it makes sense?

[the remark this answers is not in the document]
Comment (dberenbaum) [reply dvclive/metrics.py:108]: As long as it's consistent with the current `next_step`, let's merge.
